include/odp/api/spec/timer.h

```diff
@@ -106,14 +106,20 @@ typedef enum {
 typedef struct {
 	/** Timeout resolution in nanoseconds. Timer pool must serve timeouts
 	 *  with this or higher resolution. The minimum valid value (highest
-	 *  resolution) is defined by timer capability 'highest_res_ns'. */
+	 *  resolution) is defined by timer resolution capability. When this
+	 *  parameter is used, set 'res_hz' to zero. */
 	uint64_t res_ns;
 
+	/** Timeout resolution in hertz. This may be used to specify the highest
+	 *  required resolution in hertz instead of nanoseconds. When this
+	 *  parameter is used, set 'res_ns' to zero. */
+	uint64_t res_hz;
+
 	/** Minimum relative timeout in nanoseconds. All requested timeouts
 	 *  will be at least this many nanoseconds after the current
 	 *  time of the timer pool. Timer set functions return an error, if too
-	 *  short timeout was requested. The value may be also less than
-	 *  'res_ns'. */
+	 *  short timeout was requested. The value may be also smaller than
```

Comment:
A potential problem came to mind with specifying timeout always in ns. By the nature of ns vs Hz you'll always have rounding errors if timeout values are unsigned integers. That's because `Hz = 1/s`. I only see 3 ways of getting out of it (in no particular order of preference):

1. Accept this and note in documentation.
2. Specify parallel set of APIs for tick-based timeout specification (100Hz res gives 100 ticks available through API).
3. Specify that "timeout" values are specified in ns if `res_ns` was used or in ticks if `res_hz` was used.

Reply:
I think it's not a problem in practice as a user accepts a resolution size inaccuracy on timeouts anyway. E.g. minimum resolution = 30MHz = 33.333ns, timeout 7 ticks = 233ns, should arrive after 7 or 8 ticks (233.3ns - 266.6ns), depending on when timer set call was made compared to current tick period. So, in this case additional inaccuracy from nsec vs tick API is 0.3ns (233.3ns vs 233ns), but resolution based inaccuracy is 33.333 ns (100x larger).

So, I would not change timeout units right now. We'd likely need to continue towards periodic timer API definition later on and can revisit this then. I think hz as timeout value makes sense with periodic timers.

Reply:
OK.